### ISSUE TITLE
Docker pull not resolving private registry address that's been set in the host file

### ISSUE THREAD

**Issue by iZnoGod722** [moby/moby#45251]
### Description

After upgrading to version 23.0+ it seems docker is ignoring the host machines dns entry that's been set in the host file.
Our company uses two proget private docker registries hosted in 2 different servers that has separate ip addresses but sharing the same dns name.
The one registry acts as our devtest registry for developers to pull and push packages to it and once sign off of those packages has completed, those packages will then get replicated to the other registry for production use.

We have a set of build agents that are being used by our developers to perform their dev/test work. On these agents we have hardcoded the dns name as well as the ip address of the devtest registry so all pull and push activities within the build agent will be pointing to the devtest registry. Without doing so, it seems by default it will point to the production registry.

This has been working all along using docker engine version 20.10.21.

Now that I've upgraded docker engine to 23.0.1, while performing docker pull/push commands it looks like it is not resolving the registry to the ip address that's been set in the host file anymore.

`docker pull company.com/imagebuild/some-internal-image`
Error I get is

`Error response from daemon: unknown: Could not resolve ProGet feed "imagebuild". Docker respository names in ProGet must include the feed as a prefix`

The **imagebuild** feed only exists in our devtest registry so the above looks to me like docker is somehow not resolving the dns name from the host file.

I've also upgraded the engine to 23.0.2 recently and it seems it's behaving the same.

Is there any changes to the internals in terms of networking between 20.10+ and 23.0+?

### Reproduce

1. Install/Upgrade to docker 23.0.1 or 23.0.2
2. On host machine add entry to host file with ip address and dns name of the private registry
3. Try perform a docker pull of an image that exists in the private registry by using the dns name of the registry

### Expected behavior

docker pull should be able to pull the image from the private registry

### docker version

```bash
Client:
 Version:           23.0.2
 API version:       1.42
 Go version:        go1.19.7
 Git commit:        569dd73
 Built:             Mon Mar 27 16:18:04 2023
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.2
  API version:      1.42 (minimum version 1.24)
  Go version:       go1.19.7
  Git commit:       219f21b
  Built:            Mon Mar 27 16:15:45 2023
  OS/Arch:          windows/amd64
  Experimental:     false
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  v2.17.2
    Path:     C:\ProgramData\Docker\cli-plugins\docker-compose.exe

Server:
 Containers: 1
  Running: 0
  Paused: 0
  Stopped: 1
 Images: 6
 Server Version: 23.0.2
 Storage Driver: windowsfilter
  Windows:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: inactive
 Default Isolation: process
 Kernel Version: 10.0 20348 (20348.1.amd64fre.fe_release.210507-1500)
 Operating System: Microsoft Windows Server Version 21H2 (OS Build 20348.1607)
 OSType: windows
 Architecture: x86_64
 CPUs: 2
 Total Memory: 7GiB
 Name: alanc-win000000
 ID: f6da5401-c9cf-4493-8aef-52f11d38b2a6
 Docker Root Dir: C:\ProgramData\docker
 Debug Mode: true
  File Descriptors: -1
  Goroutines: 24
  System Time: 2023-03-31T21:18:51.1932554Z
  EventsListeners: 0
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

### Additional Info

Note that this is related to docker for windows, installed on windows server 2022

**Comment by thaJeztah** [comment]
Hmmm... not sure where that error comes from. That almost looks like it _is_ actually hitting your registry, but for some reason the registry rejects it?


```
Could not resolve ProGet feed "imagebuild". Docker respository names in ProGet must include the feed as a prefix
```

Does it work if you run the build with BuildKit disabled?


```bash
DOCKER_BUILDKIT=0 docker build ....... (etc)
```

**Comment by thaJeztah** [comment]
@tonistiigi does this work for BuildKit? (missed this line the first time reading)

> On host machine add entry to host file with ip address and dns name of the private registry

**Comment by thaJeztah** [comment]
oh, and this is a Windows daemon 🤔 (I should really read more carefully)

**Comment by iZnoGod722** [comment]
haha all good, I know how it feels to multitask sometimes.

So I know that from 23.0 onwards buildkit is on by default.

I'm not too familiar with the Windows daemon, but from reading this docmentation from Microsoft,

https://learn.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon

Looks like there is no option for BuildKit to be turned on/off.

**Comment by iZnoGod722** [comment]
Just an update on this @thaJeztah . I was wrong in saying that this only happens in version 23.0+.

I spun up two fresh windows server 2022 VM, one with docker client/engine version 20.10.21 and the other one 20.10.24.

And confirmed the above issue exists in 20.10.24 as well.

Perhaps this issue exists post 20.10.21?

The one difference I see is when trying to pull from a non existence registry on version 20.10.21, the error message I get is this

_Error response from daemon: Get "https://someregistry/v2/": dial tcp: lookup someregistry: no such host_

On 20.10.24 I get this

_Error response from daemon: Get "https://someregistry/v2/": dial tcp: lookup someregistry **on 10.188.132.20:53**: no such host_

The one is showing the dns server address and the other one isn't.

**Comment by thaJeztah** [comment]
The difference between those errors may be the order in which the Go's DNS resolver tries to resolve. I'm not exactly sure what order is used on Windows (on Linux, it depends whether or not there binary was built statically, and in some cases depending on the Linux user land, e.g. see https://giedrius.blog/2018/11/22/etc-nsswitch-conf-and-etc-hosts-woes-with-the-alpine-and-others-docker-image-and-golang/)

On Linux, `/etc/nsswitch.conf` can also be used (for non-static binaries) to specify priority, but not sure if there's an equivalent to that on Windows 🤔

- https://github.com/golang/go/issues/10485
- https://go.dev/src/net/conf.go

Does your host's DNS configuration have a `ndots` configured?

**Comment by iZnoGod722** [comment]
Not 100% with you here sorry, not a networking guru 😅.

I'm going to be installing docker client/engine version 20.10.22 as well as 20.10.23 to pin point the exact version where this issue starts to occur.

**Comment by iZnoGod722** [comment]
So from testing it looks like this issue starts happening from version 20.10.24 onwards.

@thaJeztah when you free, is it possible to have a look at the changes made between 20.10.23 and 20.10.24?

I'll have a look at it as well.

**Comment by thaJeztah** [comment]
At a quick glance, I thing the biggest change between those versions is updating the Go runtime from go1.18 to go1.19; https://github.com/moby/moby/compare/v20.10.23...v20.10.24

Looking at the changelog; https://go.dev/doc/go1.19#net there is mention of changes in the resolver on Windows (last paragraph);

> The pure Go resolver will now use EDNS(0) to include a suggested maximum reply packet length, permitting reply packets to contain up to 1232 bytes (the previous maximum was 512). In the unlikely event that this causes problems with a local DNS resolver, setting the environment variable GODEBUG=netdns=cgo to use the cgo-based resolver should work. Please report any such problems on the issue tracker.
> 
> When a net package function or method returns an "I/O timeout" error, the error will now satisfy errors.Is(err, context.DeadlineExceeded). When a net package function returns an "operation was canceled" error, the error will now satisfy errors.Is(err, context.Canceled). These changes are intended to make it easier for code to test for cases in which a context cancellation or timeout causes a net package function or method to return an error, while preserving backward compatibility for error messages.
> 
> Resolver.PreferGo is now implemented on Windows and Plan 9. It previously only worked on Unix platforms. Combined with Dialer.Resolver and Resolver.Dial, it's now possible to write portable programs and be in control of all DNS name lookups when dialing.
> 
> The net package now has initial support for the netgo build tag on Windows. When used, the package uses the Go DNS client (as used by Resolver.PreferGo) instead of asking Windows for DNS results. The upstream DNS server it discovers from Windows may not yet be correct with complex system network configurations, however.

**Comment by thaJeztah** [comment]
The above sounds "plausible" to be causing a change; the Windows binaries look to be using the "static" compile steps, which has `netgo` enabled. From the Go release notes, I think that previously had no effect on Windows, but starting with go1.19 it will.

I have prepared a patch to change that (and remove the `netgo` build-flag on Windows, but it looks like it can be disabled at runtime using the `GODEBUG` environment variable )`GODEBUG=netdns=cgo `).

@iZnoGod722 would you be able to start the Windows daemon with that environment variable set, and check if that resolves the issue?

**Comment by thaJeztah** [comment]
- opened a draft-PR, which should fix this (if this is indeed the cause); https://github.com/moby/moby/pull/45603

**Comment by iZnoGod722** [comment]
Hey @thaJeztah I shall give that a try.

Appreciate the help!

**Comment by thaJeztah** [comment]
Oh! A colleague just pointed out to me that that environment-variable won't work to test this 😞 https://github.com/golang/go/issues/57203

So we need to build a binary from my PR and verify with that.

To double check; you add an entry to `C:\Windows\System32\drivers\etc\hosts`, correct? So (e.g.) adding `127.0.0.1 registry-1.docker.io` should make `docker pull` fail (because it can't connect to Docker Hub's registry)

**Comment by iZnoGod722** [comment]
@thaJeztah yes that's correct, basically docker pull should be using the windows dns resolver which in turn should be looking at the host file as well and not ignoring it.

**Comment by thaJeztah** [comment]
Thanks! @neersighted is currently trying various scenarios, and it looks like its indeed fully ignored for `<unclear reasons>`, but _does_ affect `ping`, and _does_ affect things running in WSL2.

**Comment by neersighted** [comment]
I can confirm that the issue is `netgo` -- the native Go resolver does not account for the Windows hostfile. As such, the easiest thing to do for now is to built without `netgo` on Windows, and hope that the runtime improves in the future.

**Comment by thaJeztah** [comment]
@iZnoGod722 we merged fixes in our build scripts which should fix this issue (those changes will  be used for upcoming patch releases).

Thank you SO MUCH for narrowing it down to that specific patch release (leading me to the Go update). This definitely was a fun one (as most things "continued to work as usual"). I also raised this issue to other teams at Docker, and will be giving some other projects a heads-up, as it's not unlikely there's other code bases being affected as well.

So the issue is due to "file based resolution" (read: etc/hosts) just ..... not being implemented yet in Go 1.19's native resolver. And a TODO (basically; "does windows support etc/host? Let's figure that out later") was added instead.

```go
// TODO(bradfitz): implement files-based
// lookup on Windows too? I guess /etc/hosts
// kinda exists on Windows. But for now, only
// do DNS.
```

That omission was not (at least not clearly) mentioned in the release notes (but someone reported an issue about it; https://github.com/golang/go/issues/57757)

Fun fact: as `localhost` is resolved using `etc/hosts`, it also means that `localhost` won't resolve, so it definitely is severely broken (more so combining that with a previously "no-op" buildtag now becoming active _and_ introducing the breaking change).

**Comment by mika-fischer** [comment]
FYI: For the same reason [NRTP](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-r2-and-2012/dn593632(v=ws.11)#introduction-to-the-nrpt) broke (see https://github.com/docker/for-win/issues/13407). I'll see if these fixes here also fix that issue and report back.

Please don't re-enable the native go resolver unless it starts supporting NRTP on Windows (which I doubt it ever will...)...
